external/kafka/src/main/scala/org/apache/spark/rdd/kafka/KafkaCluster.scala

```diff
@@ -0,0 +1,313 @@
+    getConsumerOffsetMetadata(groupId, topicAndPartitions).right.map { r =>
+      r.map { kv =>
+        kv._1 -> kv._2.offset
+      }
+    }
+  }
+
+  def getConsumerOffsetMetadata(
+    groupId: String,
+    topicAndPartitions: Set[TopicAndPartition]
+  ): Either[Err, Map[TopicAndPartition, OffsetMetadataAndError]] = {
+    var result = Map[TopicAndPartition, OffsetMetadataAndError]()
+    val req = OffsetFetchRequest(groupId, topicAndPartitions.toSeq)
+    val errs = new Err
+    withBrokers(seedBrokers, errs) { consumer =>
+      val resp = consumer.fetchOffsets(req)
+      val respMap = resp.requestInfo
+      val needed = topicAndPartitions.diff(result.keys.toSet)
+      needed.foreach { tp =>
+        respMap.get(tp).foreach { offsetMeta =>
+          if (offsetMeta.error == ErrorMapping.NoError) {
+            result += tp -> offsetMeta
+          } else {
+            errs.append(ErrorMapping.exceptionFor(offsetMeta.error))
+          }
+        }
+      }
+      if (result.keys.size == topicAndPartitions.size) {
+        return Right(result)
+      }
+    }
+    val missing = topicAndPartitions.diff(result.keys.toSet)
+    errs.append(new Exception(s"Couldn't find consumer offsets for ${missing}"))
+    Left(errs)
+  }
+
+  def setConsumerOffsets(
+    groupId: String,
+    offsets: Map[TopicAndPartition, Long]
+  ): Either[Err, Map[TopicAndPartition, Short]] = {
+    setConsumerOffsetMetadata(groupId, offsets.map { kv =>
+      kv._1 -> OffsetMetadataAndError(kv._2)
+    })
+  }
+
+  def setConsumerOffsetMetadata(
+    groupId: String,
+    metadata: Map[TopicAndPartition, OffsetMetadataAndError]
+  ): Either[Err, Map[TopicAndPartition, Short]] = {
+    var result = Map[TopicAndPartition, Short]()
+    val req = OffsetCommitRequest(groupId, metadata)
+    val errs = new Err
+    val topicAndPartitions = metadata.keys.toSet
+    withBrokers(seedBrokers, errs) { consumer =>
+      val resp = consumer.commitOffsets(req)
+      val respMap = resp.requestInfo
+      val needed = topicAndPartitions.diff(result.keys.toSet)
+      needed.foreach { tp =>
+        respMap.get(tp).foreach { err =>
+          if (err == ErrorMapping.NoError) {
+            result += tp -> err
+          } else {
+            errs.append(ErrorMapping.exceptionFor(err))
+          }
+        }
+      }
+      if (result.keys.size == topicAndPartitions.size) {
+        return Right(result)
+      }
+    }
+    val missing = topicAndPartitions.diff(result.keys.toSet)
+    errs.append(new Exception(s"Couldn't set offsets for ${missing}"))
+    Left(errs)
+  }
+
+  private def withBrokers(brokers: Iterable[(String, Int)], errs: Err)
+    (fn: SimpleConsumer => Any): Unit = {
+    brokers.foreach { hp =>
+      var consumer: SimpleConsumer = null
+      try {
+        consumer = connect(hp)
+        fn(consumer)
+      } catch {
+        case NonFatal(e) =>
+          errs.append(e)
+      } finally {
+        if (consumer != null) consumer.close()
+      }
```

Comment:
I appreciate the feedback, but it's impossible to "stay away from java nulls" in a jvm language, without runtime checks.

Despite propaganda to the contrary, option.map is not a replacement for null checks.

The code you wrote can still throw a null pointer exception (if SimpleConsumer returns null, for instance).

You can hide the null pointer check inside of Option.apply instead of using Some, but at that point I'd rather just be explicit about what is going on.